### PR TITLE
Move EE Shared Account Types Into `casper_types`.

### LINKED ISSUE #1788: Move account types from EE to casper_types
The following types should be moved:
* `Account`
* `ActionThresholds`
* `AssociatedKeys`

### PR DESCRIPTION
**Pull request by nilz3ro**
# About These Changes
> This PR depends on #1793.
* moved `Account`, `ActionThresholds` and `AssociatedKeys` types and implementations to `casper_types::account` module.
* closes #1788.

### CODE REVIEW

**Review by Fraser999 — CHANGES_REQUESTED** [review]
The instances of `/// TODO: doc comment.` should be addressed.  Please liaise with @mpapierski on this.

**Review by rafal-ch — APPROVED** [review]
LGTM 👍

**Comment by mpapierski** [comment]
> 
> 
> The instances of `/// TODO: doc comment.` should be addressed. Please liaise with @mpapierski on this.

@nilz3ro @Fraser999 Those items you moved out from a `shared` module into `casper_types` were ticketed for documenting as part of #1765. It should be possible for you to start document these in this PR, and as far as I'm concerned there will be no collision there (I'm documenting EE's core, and Dan started EE's storage IIRC). I'm sure you'll be able to move far, and we can sync up tomorrow and I can help you with remaining items.

**Review by mpapierski — APPROVED** [review]

**Comment by mpapierski** [comment]
I think you should move last bits from https://github.com/nilz3ro/casper-node/blob/GH-1788/execution_engine/src/shared/account.rs and move it into types crate

File: execution_engine/src/shared/transform.rs [hunk 1]
```diff
@@ -1,11 +1,12 @@
 use std::{
     any,
-    convert::TryFrom,
     default::Default,
     fmt::{self, Display, Formatter},
     ops::{Add, AddAssign},
 };
 
+use core::convert::TryFrom;
+
```
Comment (Fraser999) [inline execution_engine/src/shared/transform.rs]: Was this change necessary?

File: types/src/account.rs [hunk 2]
```diff
@@ -16,11 +23,264 @@ use thiserror::Error;
...
+        Account {
+            account_hash,
+            named_keys,
+            main_purse,
+            associated_keys,
+            action_thresholds,
+        }
+    }
+
+    /// An Account constructor with presets for associated_keys and action_thresholds.
```
Comment (Fraser999) [inline types/src/account.rs:61]: I think we should describe what the presets are - not necessarily in this first paragraph, but as a further paragraph.

File: types/src/account.rs [hunk 3]
```diff
@@ -16,11 +23,264 @@ use thiserror::Error;
...
+    pub fn account_hash(&self) -> AccountHash {
+        self.account_hash
+    }
+
+    /// Returns main purse.
+    pub fn main_purse(&self) -> URef {
+        self.main_purse
+    }
+
+    /// Returns an [`AccessRights::ADD`]-only version of the [`URef`].
```
Comment (Fraser999) [inline types/src/account.rs]:
```suggestion
    /// Returns an [`AccessRights::ADD`]-only version of the main purse's [`URef`].
```

File: types/src/account.rs [hunk 4]
```diff
@@ -16,11 +23,264 @@ use thiserror::Error;
...
+
+        // Returns true if the new weight would be greater or equal to all of
+        // the thresholds.
+        new_weight >= self.action_thresholds().deployment().value()
+            && new_weight >= self.action_thresholds().key_management().value()
+    }
+
+    /// Removes an associated key from an account.
+    ///
+    /// Verifies that removing the key will not cause the remaining weight to fall below any action
```
Comment (Fraser999) [inline types/src/account.rs:158]: Maybe rephrase to something like "Returns an error if removing the key will cause the remaining weight ..."

I think an example of how thresholds work would be useful, particularly wrt to changing them.  Probably something in the top-level of this module.  The we could link to that from here and from the docs for `update_associated_key` and any others as relevant.

File: types/src/account.rs [hunk 5]
```diff
@@ -16,11 +23,264 @@ use thiserror::Error;
...
+                return Err(RemoveKeyFailure::ThresholdViolation);
+            }
+        }
+        self.associated_keys.remove_key(&account_hash)
+    }
+
+    /// Updates an associated key.
+    ///
+    /// This method validates that the key that's being updated does not have a lower weight than
+    /// the key management threshold.
```
Comment (Fraser999) [inline types/src/account.rs]: I'd rephrase to something like "Returns an error if the update would result in a violation of the key management thresholds."

Similar point for `set_action_threshold`.

File: types/src/account.rs [hunk 6]
```diff
@@ -16,11 +23,264 @@ use thiserror::Error;
...
+        weight: Weight,
+    ) -> Result<(), SetThresholdFailure> {
+        // Verify if new threshold weight exceeds total weight of all associated
+        // keys.
+        self.can_set_threshold(weight)?;
+        // Set new weight for given action
+        self.action_thresholds.set_threshold(action_type, weight)
+    }
+
+    /// Verifies if user can set action threshold
```
Comment (Fraser999) [inline types/src/account.rs]:
```suggestion
    /// Verifies if user can set action threshold.
```

File: types/src/account.rs [hunk 7]
```diff
@@ -16,11 +23,264 @@ use thiserror::Error;
...
+    pub fn can_set_threshold(&self, new_threshold: Weight) -> Result<(), SetThresholdFailure> {
+        let total_weight = self.associated_keys.total_keys_weight();
+        if new_threshold > total_weight {
+            return Err(SetThresholdFailure::InsufficientTotalWeight);
+        }
+        Ok(())
+    }
+
+    /// Checks whether all authorization keys are associated with this account
```
Comment (Fraser999) [inline types/src/account.rs]:
```suggestion
    /// Checks whether all authorization keys are associated with this account.
```

**Review by Fraser999 — COMMENTED** [review]
Nitpicks only now.

File: execution_engine/src/shared/transform.rs (hunk 1 above)
Comment (nilz3ro) [reply execution_engine/src/shared/transform.rs]: No, I will undo this change.

File: types/src/account.rs (hunk 4 above)
Comment (nilz3ro) [reply types/src/account.rs:158]: I think a practical example with doctests would be great. Do you think the examples should be done as part of this PR or should we make a new issue to add some practical examples with doc tests?

**Review by Fraser999 — APPROVED** [review]

File: types/src/account.rs (hunk 4 above)
Comment (nilz3ro) [reply types/src/account.rs:158]: Note for posterity: added #1905
